src/lib/IDBDAO.js

```diff
@@ -0,0 +1,359 @@
+        // Diminishing returns after 10000 per batch
+        if ( this.q_.length == 10000 ) this.q_ = undefined;
+      }
+    },
+
+    function withStore_(mode, fn) {
+      var self = this;
+      if ( self.__TXN__[0] ) {
+        try {
+          fn.call(self, self.__TXN__[0]);
+          return;
+        } catch (x) {
+          self.__TXN__[0] = undefined;
+        }
+      }
+      self.openDB(function (db) { // TODO: memoize like in foam1
+        var tx = db.transaction([self.name], mode);
+        var os = tx.objectStore(self.name);
+        self.__TXN__[0] = os;
+        fn.call(self, os);
+      });
+    },
+
+    function put(value) {
+      var self = this;
+      return new Promise(function(resolve, reject) {
+        self.withStore("readwrite", function(store) {
+          var request = store.put(self.serialize(value), value.id);
+          request.transaction.addEventListener(
+            'complete',
+            function(e) {
+              self.pub('on','put', value);
+              resolve(value);
+            });
+          request.transaction.addEventListener(
+            'error',
+            function(e) {
+              reject(self.IDBException.create({ id: value.id, error: e }));
+            });
+        });
+      });
+    },
+
+    function find(key) {
+      var self = this;
+
+      if ( foam.mlang.predicate.Expr.isInstance(key) ) {
+        var found = false;
+        return new Promise(function(resolve, reject) {
+          self.limit(1).where(key).select({
+            put: function(obj) {
+              found = true;
+              resolve(obj);
+            },
+            eof: function() {
+              if ( ! found ) {
+                reject(self.ObjectNotFoundException.create({ id: key }));
+              }
+            },
+          });
+        });
+      } else {
+        return new Promise(function(resolve, reject) {
+          self.withStore("readonly", function(store) {
+            var request = store.get(key);
+            request.transaction.addEventListener(
+              'complete',
+              function() {
+                if (!request.result) {
+                  reject(self.ObjectNotFoundException.create({ id: key }));
+                  return;
+                }
+                var result = self.deserialize(request.result);
+                resolve(result);
+              });
+            request.onerror = function(e) {
+              // TODO: Parse a better error out of e
+              reject(e); // TODO: err message
+            };
+          });
+        });
+      }
+    },
+
+    function remove(obj) {
+      var self = this;
+      var key = obj.id != undefined ? obj.id : obj;
+      return new Promise(function(resolve, reject) {
+        self.withStore("readwrite", function(store) {
+          var getRequest = store.get(key);
+          getRequest.onsuccess = function(e) {
+            if (!getRequest.result) {
+              resolve(null); // not found? as good as removed!
```

Comment:
To be consistent with ArrayDAO resolve with the obj passed in to remove().  We can discuss if this is correct at next design meeting.